lib/rmg/ObjectConfig.cpp

```diff
@@ -61,113 +61,219 @@ void ObjectConfig::serializeJson(JsonSerializeFormat & handler)
+						objStruct->serializeBool(subtypeKey, banned);
+					}
 				}
-			);
+			}
 		}
 	}
 	else
 	{
-		for(const auto & node : configBannedObjects.Struct())
+		const JsonNode & config = handler.getCurrent();
+		const JsonNode & configBannedObjects = config["bannedObjects"];
+
+		if(configBannedObjects.isVector())
 		{
-			LIBRARY->identifiers()->requestIdentifierIfFound(node.second.getModScope(), "object", node.first, [this, node](int primaryID)
+			// MOD COMPATIBILITY - 1.6 format
+
+			for(const auto & node : configBannedObjects.Vector())
 			{
-				if (node.second.isBool())
-				{
-					if (node.second.Bool())
-						addBannedObject(CompoundMapObjectID(primaryID, -1));
-				}
-				else
+				LIBRARY->objtypeh->resolveObjectCompoundId(node.String(),
+					[this](CompoundMapObjectID objid)
+					{
+						addBannedObject(objid);
+					}
+				);
+			}
+		}
+		else
+		{
+			for(const auto & node : configBannedObjects.Struct())
+			{
+				LIBRARY->identifiers()->requestIdentifierIfFound(node.second.getModScope(), "object", node.first, [this, node](int primaryID)
 				{
-					for (const auto & subNode : node.second.Struct())
+					if (node.second.isBool())
 					{
-						const std::string jsonKey = LIBRARY->objtypeh->getJsonKey(primaryID);
-
-						LIBRARY->identifiers()->requestIdentifierIfFound(node.second.getModScope(), jsonKey, subNode.first, [this, primaryID](int secondaryID)
+						if (node.second.Bool())
+							addBannedObject(CompoundMapObjectID(primaryID, -1));
+					}
+					else
+					{
+						for (const auto & subNode : node.second.Struct())
 						{
-							addBannedObject(CompoundMapObjectID(primaryID, secondaryID));
-						});
+							const std::string jsonKey = LIBRARY->objtypeh->getJsonKey(primaryID);
+
+							LIBRARY->identifiers()->requestIdentifierIfFound(node.second.getModScope(), jsonKey, subNode.first, [this, primaryID](int secondaryID)
+							{
+								addBannedObject(CompoundMapObjectID(primaryID, secondaryID));
+							});
+						}
 					}
-				}
-			});
+				});
+			}
 		}
 	}
 
-	for (const auto & objectConfig : configCommonObjects.Vector())
+	// Serialize common objects
+	if (handler.saving)
 	{
-		auto rmg = objectConfig["rmg"].Struct();
+		auto commonObjectsArray = handler.enterArray("commonObjects");
+		commonObjectsArray.syncSize(customObjects, JsonNode::JsonType::DATA_STRUCT);
+		
+		for(size_t i = 0; i < customObjects.size(); ++i)
+		{
+			auto objectStruct = commonObjectsArray.enterStruct(i);
+			const auto & object = customObjects[i];
+			
+			// Serialize object type/subtype
+			std::string objectType = LIBRARY->objtypeh->getJsonKey(object.primaryID);
+			objectStruct->serializeString("type", objectType);
+			
+			if(object.secondaryID != 0)
+			{
+				auto handler = LIBRARY->objtypeh->getHandlerFor(MapObjectID(object.primaryID), MapObjectSubID(object.secondaryID));
+				std::string subtypeName = handler->getSubTypeName();
+				objectStruct->serializeString("subtype", subtypeName);
+			}
```

Comment:
This condition should be removed - it might break if object has multiple subtypes. Block is correct, but must be executed unconditionally
```suggestion
			auto handler = LIBRARY->objtypeh->getHandlerFor(MapObjectID(object.primaryID), MapObjectSubID(object.secondaryID));
			std::string subtypeName = handler->getSubTypeName();
			objectStruct->serializeString("subtype", subtypeName);
```